compiler/next/lib/resolution/intents.cpp

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2021-2022 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "chpl/resolution/intents.h"
+
+#include "chpl/types/ClassType.h"
+#include "chpl/types/RecordType.h"
+#include "chpl/types/UnionType.h"
+#include "chpl/types/TupleType.h"
+
+namespace chpl {
+namespace resolution {
+
+
+using namespace uast;
+using namespace types;
+
+static QualifiedType::Kind constIntentForType(const Type* t) {
+
+  // anything we don't know the type of has to have unknown intent
+  if (t == nullptr || t->isUnknownType() || t->isErroneousType())
+    return QualifiedType::UNKNOWN;
+
+  if (t->isPrimitiveType())
+    return QualifiedType::CONST_IN;
+
+  if (t->isStringType() || t->isBytesType() ||
+      t->isRecordType() || t->isUnionType() || t->isTupleType())
+    return QualifiedType::CONST_REF;
+
+  if (auto ct = t->toClassType()) {
+    if (ct->decorator().isUnknownManagement())
+      return QualifiedType::CONST_INTENT;
+    else if (ct->decorator().isManaged())
+      return QualifiedType::CONST_REF;
+    else
+      return QualifiedType::CONST_IN;
+  }
+
+  // Otherwise, it should be a generic type that we will
+  // instantiate before computing the final intent.
+  assert(t->genericity() != Type::CONCRETE);
+  return QualifiedType::CONST_INTENT; // leave the intent generic
+}
+
+static QualifiedType::Kind defaultIntentForType(const Type* t) {
+
+  // anything we don't know the type of has to have unknown intent
+  if (t == nullptr || t->isUnknownType() || t->isErroneousType())
+    return QualifiedType::UNKNOWN;
+
+  if (t->isPrimitiveType())
+    return QualifiedType::CONST_IN;
+
+  if (t->isStringType() || t->isBytesType() ||
+      t->isRecordType() || t->isUnionType() || t->isTupleType())
+    return QualifiedType::CONST_REF;
+
+  if (auto ct = t->toClassType()) {
+    if (ct->decorator().isUnknownManagement())
+      return QualifiedType::DEFAULT_INTENT;
+    else if (ct->decorator().isManaged())
+      return QualifiedType::CONST_REF;
+    else
+      return QualifiedType::CONST_IN;
+  }
+
+  // Otherwise, it should be a generic type that we will
+  // instantiate before computing the final intent.
+  assert(t->genericity() != Type::CONCRETE);
+  return QualifiedType::DEFAULT_INTENT; // leave the intent generic
+}
+
```

Comment:
Re resolveIntent() below: the resolved intent also depends on whether it is the `this` formal. OK to keep the code as-is for now, however I think we will need to force the caller to pass an extra `isThis` argument into it.

Also I am not seeing the REF_MAYBE_CONST intent. I assume that it will be added later on.

Please add at least brief comments to the involved functions to address these two points, ex. "to be added later".